cm_tools.module

```diff
@@ -328,41 +328,24 @@ function cm_tools_array_remove_values(&$haystack, $values) {
  * @param callable $cmp_function
  */
 function cm_tools_stable_usort(&$array, $cmp_function) {
-  // Arrays of size < 2 require no action.
-  if (count($array) < 2) return;
-  // Split the array in half
-  $halfway = count($array) / 2;
-  $array1 = array_slice($array, 0, $halfway);
-  $array2 = array_slice($array, $halfway);
-  // Recurse to sort the two halves
-  cm_tools_stable_usort($array1, $cmp_function);
-  cm_tools_stable_usort($array2, $cmp_function);
-  // If all of $array1 is <= all of $array2, just append them.
-  if (call_user_func($cmp_function, end($array1), $array2[0]) < 1) {
-    $array = array_merge($array1, $array2);
-    return;
-  }
-  // Merge the two sorted arrays into a single sorted array
-  $array = array();
-  $ptr1 = $ptr2 = 0;
-  while ($ptr1 < count($array1) && $ptr2 < count($array2)) {
-    if (call_user_func($cmp_function, $array1[$ptr1], $array2[$ptr2]) < 1) {
-      $array[] = $array1[$ptr1++];
-    }
-    else {
-      $array[] = $array2[$ptr2++];
-    }
-  }
-  // Merge the remainder
-  while ($ptr1 < count($array1)) $array[] = $array1[$ptr1++];
-  while ($ptr2 < count($array2)) $array[] = $array2[$ptr2++];
-  return;
+  $index = 0;
+  foreach ($array as &$item) {
+    $item = array($index++, $item);
+  }
+  $result = usort($array, function($a, $b) use($cmp_function) {
+    $result = call_user_func($cmp_function, $a[1], $b[1]);
+    return $result == 0 ? $a[0] - $b[0] : $result;
+  });
+  foreach ($array as &$item) {
+    $item = $item[1];
+  }
```

Comment:
This is such a simple and clever way of doing this. Three cheers for whoever that guy was.